p2p/peerexchange/discovery.go

```diff
@@ -103,10 +105,15 @@ func New(logger log.Log, h host.Host, config Config) (*Discovery, error) {
 			return nil, err
 		}
 	}
-	d.scanPeers(ctx)
 	return d, nil
 }
 
+// StartScan starts background goroutine to connect with known peers
+// and scan for new ones.
+func (d *Discovery) StartScan() {
+	d.scanPeers(d.ctx)
```

Comment:
Instead of passing the context to the constructor, imo it should be passed to `StartScan` instead.

Reply:
lifecycle is controlled by Start/Stop . passing context to contstructor is not a good pattern if you have Start/Stop

Reply:
yes, that's what I said 🙂

Reply:
context should not be used to control the lifecycle of an object. Instead it should only be used to be able to cancel the function it is passed into. `StartScan` breaks this pattern which is valid for all other methods of `Discovery`.

Reply:
what I'm trying to say is: canceling the context should not be equivalent to calling `Stop` instead it should just cancel the starting